extensions/amp-analytics/0.1/resource-timing.js

```diff
@@ -0,0 +1,255 @@
+ * following properties match.
+ * @property {string=} host A string whose value should be a RegExp. It defines
+ *     a host or set of hosts to match against. By default, the RegExp will
+ *     match all hosts if omitted.
+ * @property {string=} path A string whose value should be a RegExp. It defines
+ *     a path or set of paths to match against. By default, the RegExp will
+ *     match all paths if omitted.
+ * @property {string=} query A string whose value should be a RegExp. It defines
+ *     a query string or set of query strings to match against. By default, the
+ *     RegExp will match all query strings if omitted.
+ * @typedef {!JsonObject}
+ */
+let IndividualResourceSpecDef;
+
+/**
+ * A parsed resource spec for a specific host or sets of hosts (as defined by
+ * the hostPattern).
+ * @typedef{{
+ *   hostPattern: !RegExp,
+ *   resouces: !Array<{
+ *     name: string,
+ *     pathPattern: !RegExp,
+ *     queryPattern: !RegExp,
+ *   }>,
+ * }}
+ */
+let ResourceSpecForHostDef;
+
+/**
+ * Yields the thread before running the function to avoid causing jank. (i.e. a
+ * task that takes over 16ms.)
+ * @param {function(): OUT} fn
+ * @return {!Promise<OUT>}
+ * @template OUT
+ */
+function yieldThread(fn) {
+  return new Promise(resolve => {
+    setTimeout(() => resolve(fn()));
+  });
+}
+
+/**
+ * Checks whether the given object is a valid resource timing spec.
+ * @param {!JsonObject} spec
+ * @return {boolean}
+ */
+function validateResourceTimingSpec(spec) {
+  if (!isObject(spec['resources'])) {
+    dev().warn('ANALYTICS', 'resourceTimingSpec missing "resources" field');
+    return false;
+  }
+  if (!spec['encoding'] || !spec['encoding']['entry'] ||
+      !spec['encoding']['delim']) {
+    dev().warn(
+        'ANALYTICS',
+        'resourceTimingSpec is missing or has incomplete encoding options');
+    return false;
+  }
+  if (spec['encoding']['base'] < 2 || spec['encoding']['base'] > 36) {
+    dev().warn(
+        'ANALYTICS',
+        'resource timing variables only supports bases between 2 and 36');
+    return false;
+  }
+  return true;
+}
+
+/**
+ * Gets all resource timing entries from the given window.
+ * @param {!Window} win
+ * @return {!Array<!PerformanceResourceTiming>}
+ */
+function getResourceTimingEntries(win) {
+  if (!win.performance || !win.performance.getEntriesByType) {
+    return [];
+  }
+  return /** @type {!Array<!PerformanceResourceTiming>} */ (
+    win.performance.getEntriesByType('resource'));
+}
+
+/**
+ * Converts a resource timing entry to the variables for this resource.
+ * @param {!PerformanceResourceTiming} entry
+ * @param {string} name Name of the resource set by the resourceTimingSpec.
+ * @param {function(number,number=): string} format A function to format
+ *    timestamps and intervals. (Two numbers will be passed in for intervals.)
+ * @return {!ExpansionOptions}
+ */
+function entryToExpansionOptions(entry, name, format) {
```

Comment:
Agreed that it's not 100% needed, but it eliminates some code duplication. I pass it around as a closure so that it can be re-used by multiple entries.